tt_metal/distributed/mesh_program.cpp

```diff
@@ -0,0 +1,144 @@
+    MeshProgram& mesh_program,
+    const std::variant<CoreRange, CoreRangeSet> &core_spec,
+    uint32_t initial_value,
+    CoreType core_type) {
+    return mesh_program.distributed_impl_(
+        std::function<uint32_t(Program&)>(
+            [&core_spec, initial_value, core_type] (Program& program) -> uint32_t {
+                return CreateSemaphore(program, core_spec, initial_value, core_type);
+            }
+        )
+    );
+}
+
+uint32_t CreateSemaphore(
+    MeshProgram& mesh_program,
+    const std::variant<CoreRange, CoreRangeSet> &core_spec,
+    uint32_t initial_value,
+    CoreType core_type,
+    chip_id_t device_id) {
+    return CreateSemaphore(mesh_program.at(device_id), core_spec, initial_value, core_type);
+}
+
+CBHandle CreateCircularBuffer(
+    MeshProgram& mesh_program,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const CircularBufferConfig &config) {
+    return mesh_program.distributed_impl_(
+        std::function<CBHandle(Program&)>(
+            [&core_spec, &config] (Program& program) -> CBHandle {
+                return CreateCircularBuffer(program, core_spec, config);
+            }
+        )
+    );
+}
+
+CBHandle CreateCircularBuffer(
+    MeshProgram& mesh_program,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const CircularBufferConfig &config,
+    chip_id_t device_id) {
+        return CreateCircularBuffer(mesh_program.at(device_id), core_spec, config);
+}
+
+void SetRuntimeArgs(
+    MeshProgram& mesh_program,
+    KernelHandle kernel,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::vector<uint32_t> &runtime_args) {
+    mesh_program.distributed_impl_(
+        std::function<void(Program&)>(
+            [kernel, &core_spec, &runtime_args] (Program& program) -> void {
+                return SetRuntimeArgs(program, kernel, core_spec, runtime_args);
+            }
+        )
+    );
+}
+
+void SetRuntimeArgs(
+    MeshProgram& mesh_program,
+    KernelHandle kernel,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::vector<uint32_t> &runtime_args,
+    chip_id_t device_id) {
+        SetRuntimeArgs(mesh_program.at(device_id), kernel, core_spec, runtime_args);
+}
+
+KernelHandle CreateKernel(
+    MeshProgram& mesh_program,
+    const std::string &file_name,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
+    const std::variant<DataMovementConfig, ComputeConfig, EthernetConfig> &config) {
+    return mesh_program.distributed_impl_(
+        std::function<KernelHandle(Program&)>(
+            [&file_name, &core_spec, &config] (Program& program) -> KernelHandle {
+                return CreateKernel(program, file_name, core_spec, config);
+            }
+        )
+    );
+}
+
+KernelHandle CreateKernel(
+    MeshProgram& mesh_program,
+    const std::string &file_name,
+    const std::variant<CoreCoord, CoreRange, CoreRangeSet> &core_spec,
```

Comment:
It feels like we shouldn't be trying to program to the MeshProgram and then to also the Program level or we'll end up feeding two sets of inputs at different abstraction levels